node/subsystem/src/util.rs

```diff
@@ -429,10 +435,145 @@ impl<Spawner: Spawn, Job: JobTrait> Jobs<Spawner, Job> {
 impl<Spawner, Job: JobTrait> Drop for Jobs<Spawner, Job> {
 	fn drop(&mut self) {
 		for job_handle in self.running.values() {
 			job_handle.abort_handle.abort();
 		}
 	}
 }
+
+/// A basic implementation of a subsystem.
+///
+/// This struct is responsible for handling message traffic between
+/// this subsystem and the overseer. It spawns and kills jobs on the
+/// appropriate Overseer messages, and dispatches standard traffic to
+/// the appropriate job the rest of the time.
+pub struct JobManager<Spawner, Context, Job: JobTrait> {
+	spawner: Spawner,
+	run_args: Job::RunArgs,
+	context: std::marker::PhantomData<Context>,
+	job: std::marker::PhantomData<Job>,
+}
+
+impl<Spawner, Context, Job> JobManager<Spawner, Context, Job>
+where
+	Spawner: Spawn + Clone + Send,
+	Context: SubsystemContext,
+	<Context as SubsystemContext>::Message: Into<Job::ToJob>,
+	Job: JobTrait,
+	Job::RunArgs: Clone,
+	Job::ToJob: TryFrom<AllMessages> + Sync + Debug,
+{
+	/// Creates a new `Subsystem`.
+	pub fn new(spawner: Spawner, run_args: Job::RunArgs) -> Self {
+		Self {
+			spawner,
+			run_args,
+			context: std::marker::PhantomData,
+			job: std::marker::PhantomData,
+		}
+	}
+
+	/// Run this subsystem
+	///
+	/// Conceptually, this is very simple: it just loops forever.
+	///
+	/// - On incoming overseer messages, it starts or stops jobs as appropriate.
+	/// - On other incoming messages, if they can be converted into Job::ToJob and
+	///   include a hash, then they're forwarded to the appropriate individual job.
+	/// - On outgoing messages from the jobs, it forwards them to the overseer.
+	pub async fn run(mut ctx: Context, run_args: Job::RunArgs, spawner: Spawner) {
+		let mut jobs = Jobs::new(spawner.clone());
+
+		loop {
+			select! {
+				incoming = ctx.recv().fuse() => if Self::handle_incoming(incoming, &mut jobs, &run_args).await { break },
+				outgoing = jobs.next().fuse() => if Self::handle_outgoing(outgoing, &mut ctx).await { break },
+				complete => break,
+			}
+		}
+	}
+
+	// handle an incoming message. return true if we should break afterwards.
+	async fn handle_incoming(
+		incoming: SubsystemResult<FromOverseer<Context::Message>>,
+		jobs: &mut Jobs<Spawner, Job>,
+		run_args: &Job::RunArgs,
+	) -> bool {
+		use crate::FromOverseer::{Communication, Signal};
+		use crate::OverseerSignal::{Conclude, StartWork, StopWork};
+
+		match incoming {
+			Ok(Signal(StartWork(hash))) => {
+				if let Err(e) = jobs.spawn_job(hash, run_args.clone()) {
+					log::error!("Failed to spawn a job: {:?}", e);
+					return true;
+				}
+			}
+			Ok(Signal(StopWork(hash))) => {
+				if let Err(e) = jobs.stop_job(hash).await {
+					log::error!("Failed to stop a job: {:?}", e);
+					return true;
+				}
+			}
+			Ok(Signal(Conclude)) => {
+				// breaking the loop ends fn run, which drops `jobs`, which drops all ongoing work
+				return true;
+			}
+			Ok(Communication { msg }) => {
+				// I don't know the syntax to add this type hint within an if let statement
```

Comment:
I guess you can use a type annotated `try_from` like `Result<Job::ToJob, _>::try_from(msg)`. 

Although it requires `TryFrom` trait, but it should be implemented anyway, as `TryFrom` implies `TryInto` https://doc.rust-lang.org/std/convert/trait.TryFrom.html#generic-implementations, but not the other way around (same for `From` and `Into`).

But it's fine as is, no need to worry.

Reply:
Thanks, that was a good idea!